internal/allocator/allocator.go

```diff
@@ -194,113 +222,116 @@ func (a *Allocator) Unassign(svc string) bool {
+		if len(a.portsInUse[ip.String()]) == 0 {
+			delete(a.portsInUse, ip.String())
+			delete(a.sharingKeyForIP, ip.String())
+		}
+		a.poolIPsInUse[al.pool][ip.String()]--
+		if a.poolIPsInUse[al.pool][ip.String()] == 0 {
+			// Explicitly delete unused IPs from the pool, so that len()
+			// is an accurate count of IPs in use.
+			delete(a.poolIPsInUse[al.pool], ip.String())
 		}
-		delete(a.portsInUse[al.ip.String()], port)
-	}
-	delete(a.servicesOnIP[al.ip.String()], svc)
-	if len(a.portsInUse[al.ip.String()]) == 0 {
-		delete(a.portsInUse, al.ip.String())
-		delete(a.sharingKeyForIP, al.ip.String())
-	}
-	a.poolIPsInUse[al.pool][al.ip.String()]--
-	if a.poolIPsInUse[al.pool][al.ip.String()] == 0 {
-		// Explicitly delete unused IPs from the pool, so that len()
-		// is an accurate count of IPs in use.
-		delete(a.poolIPsInUse[al.pool], al.ip.String())
 	}
 	stats.poolActive.WithLabelValues(al.pool).Set(float64(len(a.poolIPsInUse[al.pool])))
 	return true
 }
 
-func cidrIsIPv6(cidr *net.IPNet) bool {
-	return cidr.IP.To4() == nil
-}
-func ipIsIPv6(ip net.IP) bool {
-	return ip.To4() == nil
-}
-
 // AllocateFromPool assigns an available IP from pool to service.
-func (a *Allocator) AllocateFromPool(svc string, isIPv6 bool, poolName string, ports []Port, sharingKey, backendKey string) (net.IP, error) {
+func (a *Allocator) AllocateFromPool(svc string, ipFamily IPFamily, poolName string, ports []Port, sharingKey, backendKey string) ([]net.IP, error) {
 	if alloc := a.allocated[svc]; alloc != nil {
-		// Handle the case where the svc has already been assigned an IP but from the wrong family.
-		// This "should-not-happen" since the "ipFamily" is an immutable field in services.
-		if isIPv6 != ipIsIPv6(alloc.ip) {
-			return nil, fmt.Errorf("IP for wrong family assigned %s", alloc.ip.String())
+		if ipFamily != DualStack {
+			// Handle the case where the svc has already been assigned an IP but from the wrong family.
+			// This "should-not-happen" since the "ipFamily" is an immutable field in services.
+			if ipFamily != getIPFamily(alloc.ips[0]) {
+				return nil, fmt.Errorf("IP for wrong family assigned %s", alloc.ips[0].String())
+			}
 		}
-		if err := a.Assign(svc, alloc.ip, ports, sharingKey, backendKey); err != nil {
+		if err := a.Assign(svc, alloc.ips, ports, sharingKey, backendKey); err != nil {
 			return nil, err
 		}
-		return alloc.ip, nil
+		return alloc.ips, nil
 	}
 
 	pool := a.pools[poolName]
 	if pool == nil {
 		return nil, fmt.Errorf("unknown pool %q", poolName)
 	}
 
+	sk := &key{
+		sharing: sharingKey,
+		backend: backendKey,
+	}
+	ips := []net.IP{}
 	for _, cidr := range pool.CIDR {
-		if cidrIsIPv6(cidr) != isIPv6 {
-			// Not the right ip-family
-			continue
-		}
-		c := ipaddr.NewCursor([]ipaddr.Prefix{*ipaddr.NewPrefix(cidr)})
-		for pos := c.First(); pos != nil; pos = c.Next() {
-			ip := pos.IP
-			if pool.AvoidBuggyIPs && ipConfusesBuggyFirmwares(ip) {
+		if len(ips) == 0 {
```

Comment:
I think this is not working in the dual stack case.
What we need to do here is:
- find a v4 ip, a v6 ip (if single stack).
- find a v4 AND a v6 from the pool if dual stack.

Here we always check ips[0] and we never stop, so ips will grow more than two items if the pool has many cidrs.

You can do something like:

```
ips := []net.IP{}
	v4ToAdd := true
	v6ToAdd := true

	switch family {
	case ipfamily.IPV4:
		v6ToAdd = false
	case ipfamily.IPV6:
		v4ToAdd = false
	}

	for _, cidr := range pool.CIDR {
		if cidrIsIPv6(cidr) && !v6ToAdd {
			continue
		}
		if !cidrIsIPv6(cidr) && !v4ToAdd {
			continue
		}

		c := ipaddr.NewCursor([]ipaddr.Prefix{*ipaddr.NewPrefix(cidr)})
		for pos := c.First(); pos != nil; pos = c.Next() {
			ip := pos.IP
			if pool.AvoidBuggyIPs && ipConfusesBuggyFirmwares(ip) {
				continue
			}
			// Somewhat inefficiently brute-force by invoking the
			// IP-specific allocator.
			....
		}

	}
```

Reply:
I see that is much cleaner logic